gptfixer/gpt.c

```diff
@@ -0,0 +1,786 @@
+        errx(1, "Cannot fit new partition table into disk space");
+    }
+    // This is the largest value FirstUsableLBA can be set to.
+    uint64_t first_start_byte_offset = UINT64_MAX;
+    uint64_t last_end_byte_offset = 0;
+    uint32_t partition_that_starts_first = UINT32_MAX;
+    uint32_t partition_that_ends_last = UINT32_MAX;
+    for_each_gpt_entry (i, size, count, gpt, entry) {
+        if (unused_entry(entry)) {
+            continue;
+        }
+        uint64_t start_byte_offset = entry->StartingLBA * old_sector_size;
+        uint64_t end_byte_offset = (entry->EndingLBA + 1) * old_sector_size;
+        vlog("Partition consumes bytes [0x%" PRIX64 ", 0x%" PRIX64 ")", start_byte_offset, end_byte_offset);
+        if (start_byte_offset % sector_size != 0) {
+            errx(1,
+                 "Cannot preserve starting LBA for partition %" PRIu32 ": starting byte offset 0x%" PRIX64
+                 " is not a multiple of %" PRIu32,
+                 i + 1, start_byte_offset, sector_size);
+        }
+        if (end_byte_offset % sector_size != 0) {
+            // TODO better error message
+            errx(1,
+                 "Cannot preserve ending LBA for partition %" PRIu32 ": ending byte offset 0x%" PRIX64
+                 " is not a multiple of %" PRIu32,
+                 i + 1, end_byte_offset, sector_size);
+        }
+        if (start_byte_offset < first_start_byte_offset) {
+            first_start_byte_offset = start_byte_offset;
+            partition_that_starts_first = i;
+        }
+        if (last_end_byte_offset < end_byte_offset) {
+            last_end_byte_offset = end_byte_offset;
+            partition_that_ends_last = i;
+        }
+    }
+
+    // Add sector_size twice: once for protective MBR, and once for
+    // primary GPT header.  Use 64-bit multiplication to prevent wraparound
+    // in the absurd case of 2**31-byte sectors.
+    uint64_t const end_of_primary_gpt_table = bytes_required + UINT64_C(2) * sector_size;
+    if (first_start_byte_offset < end_of_primary_gpt_table) {
+        errx(1,
+             "Cannot construct new GPT: Partition %" PRIu32 " starts at byte offset 0x%" PRIX64
+             ", but partition table uses space up to 0x%" PRIX64,
+             partition_that_starts_first + 1, first_start_byte_offset, end_of_primary_gpt_table);
+    }
+
+    // Add sector_size once for backup GPT header.
+    uint64_t start_of_backup_gpt_table = size - (bytes_required + sector_size);
+    if (last_end_byte_offset > start_of_backup_gpt_table) {
+        errx(1,
+             "Cannot construct new GPT: Partition %" PRIu32 " ends at byte offset 0x%" PRIX64
+             ", but partition table uses space after 0x%" PRIX64,
+             partition_that_ends_last + 1, last_end_byte_offset, start_of_backup_gpt_table);
+    }
+    uint64_t first_usable_byte;
+    CHECK(!__builtin_mul_overflow(gpt->header.FirstUsableLBA, old_sector_size, &first_usable_byte));
```

Comment:
Technically, this should be proper error check, with a message about too large FirstUsableLBA value (aka invalid GPT). Alternatively, you could check FirstUsableLBA against the number of sectors on the disk - to get more meaningful error message, but I don't think you have `old_sectors` anywhere already?